src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc

```diff
@@ -23,146 +23,188 @@ namespace ray {
+    const ResourceRequest &resource_request,
+    float spread_threshold,
+    bool force_spillback,
+    bool require_node_available,
+    NodeFilter node_filter,
+    int32_t schedule_top_k_absolute,
+    float scheduler_top_k_fraction) {
+  // Nodes that are feasible and currently have available resources.
+  std::vector<std::pair<scheduling::NodeID, float>> available_nodes;
+  // Nodes that are feasible but currently do not have available resources.
+  std::vector<std::pair<scheduling::NodeID, float>> feasible_nodes;
+  // Check whether the local node is available and feasible. We'll use this to
+  // help prioritize the local node when force_spillback=false.
+  bool local_node_is_available = false;
+  bool local_node_is_feasible = false;
+  for (const auto &pair : nodes_) {
+    const auto &node_id = pair.first;
+    const auto &node_resources = pair.second.GetLocalView();
+    if (force_spillback && node_id == local_node_id_) {
+      continue;
+    }
+    if (IsNodeFeasible(node_id, node_filter, node_resources, resource_request)) {
+      bool ignore_pull_manager_at_capacity = false;
+      if (node_id == local_node_id_) {
+        // It's okay if the local node's pull manager is at
+        // capacity because we will eventually spill the task
+        // back from the waiting queue if its args cannot be
+        // pulled.
+        ignore_pull_manager_at_capacity = true;
+        local_node_is_feasible = true;
+      }
+      bool is_available =
+          node_resources.IsAvailable(resource_request, ignore_pull_manager_at_capacity);
+      if (node_id == local_node_id_ and is_available) {
+        local_node_is_available = true;
+      }
+      float node_score = ComputeNodeScore(node_resources, spread_threshold);
+      RAY_LOG(DEBUG) << "Node " << node_id.ToInt() << " is "
+                     << (is_available ? "available" : "not available") << " for request "
+                     << resource_request.DebugString()
+                     << " with critical resource utilization " << node_score
+                     << " based on local view " << node_resources.DebugString();
+      if (is_available) {
+        available_nodes.push_back({node_id, node_score});
+      } else {
+        feasible_nodes.push_back({node_id, node_score});
+      }
     }
   }
 
-  return best_node_id;
+  size_t num_candidate_nodes = std::max<int32_t>(
+      1,
+      std::max<int32_t>(schedule_top_k_absolute,
+                        static_cast<int32_t>(nodes_.size() * scheduler_top_k_fraction)));
+
+  if (!available_nodes.empty()) {
+    bool prioritize_local_node = !force_spillback && local_node_is_available;
+    // First prioritize available nodes.
+    return GetBestNode(
+        available_nodes,
+        num_candidate_nodes,
+        prioritize_local_node,
+        prioritize_local_node ? CompuateLocalNodeScore(spread_threshold) : 1);
+  } else if (!feasible_nodes.empty() && !require_node_available) {
```

Comment:
what is `require_node_available` for?

Reply:
can you add comment on this bool why this condition means to "prioritize_local_node"?